test/Concurrency/Runtime/async_sequence.swift

```diff
@@ -0,0 +1,524 @@
+// and async and throwing async contexts.
+
+func throwing<T>(_ value: T) throws -> T {
+  return value
+}
+
+func asynchronous<T>(_ value: T) async -> T {
+  return value
+}
+
+func asynchronousThrowing<T>(_ value: T) async throws -> T {
+  return value
+}
+
+struct Failure: Error { }
+
+func failable<T, E: Error>(
+  _ results: [Result<T, E>]
+) -> AsyncThrowingMapSequence<AsyncLazySequence<[Result<T, E>]>, T> {
+  return results.async.map { try $0.get() }
+}
+
+
+extension Sequence {
+  @inlinable
+  public var async: AsyncLazySequence<Self> {
+    get {
+      return AsyncLazySequence(self)
+    }
+  }
+}
+
+@frozen
+public struct AsyncLazySequence<S: Sequence>: AsyncSequence {
+  public typealias Element = S.Element
+  public typealias AsyncIterator = Iterator
+
+  @frozen
+  public struct Iterator: AsyncIteratorProtocol {
+    @usableFromInline
+    var iterator: S.Iterator
+
+    @usableFromInline
+    init(_ iterator: S.Iterator) {
+      self.iterator = iterator
+    }
+
+    @inlinable
+    public mutating func next() async -> S.Element? {
+      return iterator.next()
+    }
+  }
+
+  public let sequence: S
+
+  @inlinable
+  public init(_ sequence: S) {
+    self.sequence = sequence
+  }
+
+  @inlinable
+  public __consuming func makeAsyncIterator() -> Iterator {
+    return Iterator(sequence.makeIterator())
+  }
+}
+
+extension AsyncSequence {
+  @inlinable
+  public func collect() async rethrows -> [Element] {
+    var items = [Element]()
+    var it = makeAsyncIterator()
+    while let e = try await it.next() {
+      items.append(e)
+    }
+    return items
+  }
+}
+
+extension TestSuite {
+  @inline(never)
+  public func test(
+    _ name: String,
+    file: String = #file, line: UInt = #line,
+    _ testFunction: @escaping () async -> Void
+  ) {
+    test(name, file: file, line: line) {
+      runAsyncAndBlock(testFunction)
+    }
+  }
+}
+
+var AsyncLazySequenceTests = TestSuite("AsyncLazySequence")
+
+AsyncLazySequenceTests.test("iteration past first nil") {
+  let seq = [1, 2, 3]
+  let async_seq = seq.async
+  var iter = async_seq.makeAsyncIterator()
+  let a = await iter.next()
+  expectEqual(a, 1)
+  let b = await iter.next()
+  expectEqual(b, 2)
+  let c = await iter.next()
+  expectEqual(c, 3)
+  let d = await iter.next()
+  expectNil(d)
+  let e = await iter.next()
+  expectNil(e)
+}
+
+AsyncLazySequenceTests.test("iterate twice with iterators before usage") {
+  let s = [1, 2, 3].async
+  var iter1 = s.makeAsyncIterator()
+  var iter2 = s.makeAsyncIterator()
+  var iter1Results = [Int?]()
+  var iter2Results = [Int?]()
+  for _ in 1...4 {
+    do {
+      let value = await iter1.next()
+      iter1Results.append(value)
+    }
+
+    do {
+      let value = await iter2.next()
+      iter2Results.append(value)
+    }
+  }
+  expectEqual(iter1Results, iter2Results)
+  expectEqual(iter1Results, [1, 2, 3, nil])
+}
+
+var AsyncSequenceTests = TestSuite("AsyncSequence")
```

Comment:
I think we need tests that throw for all these single-value producing methods.

Reply:
added a test for each,  both in throwing with closure but also in throwing from upstreams

Reply:
also added tests for non closure async functions